File: dalesbred/src/test/java/fi/evident/dalesbred/DatabaseCustomDialectTest.java
package fi.evident.dalesbred;

import fi.evident.dalesbred.dialects.DefaultDialect;
import org.jetbrains.annotations.NotNull;
import org.junit.Rule;
import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class DatabaseCustomDialectTest {

    private final Database db = TestDatabaseProvider.createTestDatabase(new UppercaseDialect());

    @Rule
    public final TransactionalTestsRule rule = new TransactionalTestsRule(db);

    @Test
    public void customDialect() {
        db.update("drop table if exists my_table");
        db.update("create table my_table (text varchar)");

        db.update("insert into my_table values (?)", "foo");

        assertEquals("FOO", db.findUnique(String.class, "select text from my_table"));
    }

    private static final class UppercaseDialect extends DefaultDialect {
        @NotNull
        @Override
        public Object valueToDatabase(@NotNull Object value) {
            return value.toString().toUpperCase();
        }
    }
}

File: dalesbred/src/test/kotlin/org/dalesbred/DatabaseCustomDialectTest.kt
package org.dalesbred

import org.dalesbred.connection.ConnectionProvider
import org.dalesbred.dialect.DefaultDialect
import org.dalesbred.testutils.DatabaseProvider.POSTGRESQL
import org.dalesbred.testutils.DatabaseTest
import org.dalesbred.testutils.transactionalTest
import kotlin.test.Test
import kotlin.test.assertEquals

@DatabaseTest(POSTGRESQL)
class DatabaseCustomDialectTest(connectionProvider: ConnectionProvider) {

    private val db = Database(connectionProvider, UppercaseDialect)

    @Test
    fun `custom Dialect`() = transactionalTest(db) {
        db.update("drop table if exists my_table")
        db.update("create temporary table my_table (text varchar(64))")

        db.update("insert into my_table values (?)", "foo")

        assertEquals("FOO", db.findUnique(String::class.java, "select text from my_table"))
    }

    private object UppercaseDialect : DefaultDialect() {
        override fun valueToDatabase(value: Any) = value.toString().uppercase()
    }
}
